### DISCUSSION TITLE
OpenAPI/Swagger show certain routes only to logged in users

### DISCUSSION THREAD

**Discussion by lsabi** [fastapi/fastapi#8869] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description
It would be useful to show some routes or paths only to logged in users. That is, if I have routes `A, B, C` on a public endpoint, but `B` and `C` are `POST` and `PUT` routes accessible only to authenticated users, it would be a nice feature to be able to hide them and show them only to logged in users.

In some cases, just part of the information should be restricted to certain users, while other information can remain public. The case could be a blog, where everyone can request posts, but may not add a new one, unless authorized (i.e. logged in).

Another use case could be e-commerce, which I've noted, that e-commerce owners are providing a public API for getting the products and searching, while `POST`ing and `PUT`ting orders is forbidden to non registered users.

`UI`s that run almost only on the client side are more and more popular, and they use `API`s to access data. `API`s which are on a different domain or subdomain (e.g. `www.domain.com` and `www.api.domain.com`). Thus it is possible that some routes are freely accessible, while others are not. This feature request covers the routes in the latter case.

In some cases, `API` owners, may not care and let everyone know what the `API` looks like for all the routes, but I believe it would be better to hide information that you are not supposed to use because you're not authenticated.

### Example

Here's an idea of what it could look like (it's a get request, but it serves to get my point)

```Python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
@app.docsOnlyLoggedIn
def read_root():
    return {"Hello": "World"}
```

### Describe alternatives you've considered

Creating other apps with fastapi, though it would become a redundancy for most of the paths. Also, it would become really complicate to manage two apps that run on the same domain and share the prefix.

Alternatively, not caring or removing the docs, but I'm not much in favor of such a thing.

### Final Notes

Mine is just a proposal. I couldn't find docs/post/issue mentioning such a request. Also, if it's not too much of a hassle to integrate the feature, otherwise the issue can be closed. Possibly a poll could help understand if it's worth digging into or not.

This should be an issue related to `fastapi` and not swagger itself, since `fastapi` provides the routes to swagger, doesn't it?

**Comment by STeveShary** [comment]
With FastAPI, you can split your docs using sub-applications: https://fastapi.tiangolo.com/advanced/sub-applications/

You should be able to configure Oath2 for the sub application: https://fastapi.tiangolo.com/advanced/security/oauth2-scopes/

**Comment by lsabi** [comment]
@STeveShary thanks, but I have an API, of which only certain paths I want to show to not logged in users. That is, secure routes are not what I need.

Also, this way I need to have different route paths, which is not what I want. If there's the blog post at `/api/posts`, I want people to be able to see `GET` docs, but not `POST` docs and to be able to query `GET` with no auth, but to `POST` with an auth token.

**Comment by LorhanSohaky** [comment]
The sub app doesn't work properly because of this bug #4656

**Comment by YuriiMotov** [comment, marked as answer]
This question is more for Swagger UI repo.
You can try solve this using self-hosted Swagger Docs and modifying its assets
